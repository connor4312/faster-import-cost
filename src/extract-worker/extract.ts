import type { Options } from "acorn";
import { parse } from "acorn-loose";
import { traverse } from "estraverse";
import { Node } from "estree";
import { IImport } from "../types";

export const acornOptions: Options = {
  ecmaVersion: "latest",
  locations: true,
  allowAwaitOutsideFunction: true,
  allowImportExportEverywhere: true,
  allowReserved: true,
  allowReturnOutsideFunction: true,
};

export function extract(text: string) {
  const ast = parse(text, acornOptions);
  const ranges: IImport[] = [];

  traverse(ast as Node, {
    enter(node) {
      if (node.type === "ImportDeclaration") {
        ranges.push({
          range: node.loc!,
          name: String(node.source.value),
          members: node.specifiers.some((s) => s.type === "ImportDefaultSpecifier")
            ? "default"
            : node.specifiers.some((s) => s.type === "ImportNamespaceSpecifier")
            ? "all"
            : node.specifiers.map((s) => s.local.name!),
        });
      } else if (node.type === "ImportExpression" && node.source.type === "Literal") {
        ranges.push({
          range: node.loc!,
          name: String(node.source.value),
          members: "all",
        });
      }
    },
  });

  return ranges;
}
