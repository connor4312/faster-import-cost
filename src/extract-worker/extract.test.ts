import { describe, expect, it } from "vitest";
import { extract } from "./extract";

describe("extract", () => {
  it("extracts simple js", () => {
    expect(extract('import { foo } from "bar";')).toMatchSnapshot();
  });

  it("extracts default type", () => {
    expect(extract('import foo from "bar";')).toMatchSnapshot();
  });

  it("extracts namespace", () => {
    expect(extract('import * as asdf from "bar";')).toMatchSnapshot();
  });

  it("extracts inline import", () => {
    expect(extract('const bar = await import("bar");')).toMatchSnapshot();
  });
});
