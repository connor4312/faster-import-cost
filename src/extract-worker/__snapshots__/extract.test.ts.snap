// Vitest Snapshot v1

exports[`extract > extracts default type 1`] = `
[
  {
    "members": "default",
    "name": "bar",
    "range": SourceLocation {
      "end": Position {
        "column": 22,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
  },
]
`;

exports[`extract > extracts inline import 1`] = `
[
  {
    "members": "all",
    "name": "bar",
    "range": SourceLocation {
      "end": Position {
        "column": 31,
        "line": 1,
      },
      "start": Position {
        "column": 18,
        "line": 1,
      },
    },
  },
]
`;

exports[`extract > extracts namespace 1`] = `
[
  {
    "members": "all",
    "name": "bar",
    "range": SourceLocation {
      "end": Position {
        "column": 28,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
  },
]
`;

exports[`extract > extracts simple js 1`] = `
[
  {
    "members": [
      "foo",
    ],
    "name": "bar",
    "range": SourceLocation {
      "end": Position {
        "column": 26,
        "line": 1,
      },
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
  },
]
`;
